bugsnag-plugin-android-okhttp/src/main/java/com/bugsnag/android/okhttp/BugsnagOkHttpPlugin.kt

```diff
@@ -1,14 +1,148 @@
+ * call succeeds or failed.
+ *
+ * To enable this functionality in Bugsnag call [com.bugsnag.android.Configuration.addPlugin]
+ * with an instance of this object before calling [com.bugsnag.android.Bugsnag.start].
+ */
+class BugsnagOkHttpPlugin @JvmOverloads constructor(
+    internal val timeProvider: () -> Long = { System.currentTimeMillis() }
+) : Plugin, EventListener() {
+
+    internal val requestMap = ConcurrentHashMap<Call, NetworkRequestMetadata>()
+    private var client: Client? = null
 
     override fun load(client: Client) {
+        this.client = client
     }
 
     override fun unload() {
+        this.client = null
+    }
+
+    override fun callStart(call: Call) {
+        requestMap[call] = NetworkRequestMetadata(timeProvider())
+    }
+
+    override fun canceled(call: Call) {
+        requestMap.remove(call)
+    }
+
+    override fun requestBodyEnd(call: Call, byteCount: Long) {
+        requestMap[call]?.requestBodyCount = byteCount
+    }
+
+    override fun responseBodyEnd(call: Call, byteCount: Long) {
+        requestMap[call]?.responseBodyCount = byteCount
+    }
+
+    override fun responseHeadersEnd(call: Call, response: Response) {
+        requestMap[call]?.status = response.code
+    }
+
+    override fun callEnd(call: Call) {
+        requestMap.remove(call)?.let { requestInfo ->
+            client?.apply {
+                leaveBreadcrumb(
+                    "OkHttp call succeeded",
+                    collateMetadata(call, requestInfo, timeProvider()),
+                    BreadcrumbType.REQUEST
+                )
+            }
+        }
+    }
+
+    override fun callFailed(call: Call, ioe: IOException) {
+        requestMap.remove(call)?.let { requestInfo ->
+            client?.apply {
+                leaveBreadcrumb(
+                    "OkHttp call failed",
+                    collateMetadata(call, requestInfo, timeProvider()),
+                    BreadcrumbType.REQUEST
+                )
+            }
+        }
     }
+
+    @VisibleForTesting
+    internal fun Client.collateMetadata(
+        call: Call,
+        info: NetworkRequestMetadata,
+        nowMs: Long
+    ): Map<String, Any> {
+        val request = call.request()
+
+        return mapOf(
+            "method" to request.method,
+            "url" to sanitizeUrl(request),
+            "duration" to nowMs - info.startTime,
+            "urlParams" to buildQueryParams(request),
+            "requestContentLength" to info.requestBodyCount,
+            "responseContentLength" to info.responseBodyCount,
+            "status" to info.status
+        )
+    }
+
+    /**
+     * Constructs a map of query parameters, redacting any sensitive values.
+     */
+    private fun Client.buildQueryParams(request: Request): Map<String, String?> {
+        val url = request.url
+        val params = mutableMapOf<String, String?>()
+
+        url.queryParameterNames.forEach { name ->
+            params[name] = url.queryParameter(name)
```

Comment:
We're only capturing a single value for each query parameter. Certainly the 99% case is covered, but I wonder if it's not worth having some logic to handle repeated parameters?

Reply:
`queryParameterNames` is a `Set` so I think it shouldn't be possible to have duplicate keys?

Reply:
Updated so that a list is now added if more than one value is present.